scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/scheduler/engine/daemon/ReplaceAfterTerminateScheduler.java

```diff
@@ -15,18 +15,35 @@
 package com.amazonaws.blox.scheduling.scheduler.engine.daemon;
 
 import com.amazonaws.blox.scheduling.scheduler.engine.SchedulingAction;
+import com.amazonaws.blox.scheduling.state.ClusterSnapshot;
 import java.util.List;
 import java.util.stream.Collectors;
 
 public class ReplaceAfterTerminateScheduler extends DaemonScheduler {
   public static final String ID = "ReplaceAfterTerminate";
 
   public List<SchedulingAction> schedule(DaemonEnvironment env, ClusterSummary summary) {
-    return summary
-        .getInstances()
-        .stream()
-        .filter(i -> env.hasMatchingTask(summary.tasksForInstance(i)))
-        .map(i -> env.startTaskFor(i))
-        .collect(Collectors.toList());
+    List<ClusterSnapshot.Task> tasksToStop =
+        summary
+            .getInstances()
+            .stream()
+            .filter(i -> env.needsToStop(summary.tasksForInstance(i)))
+            .flatMap(i -> summary.tasksForInstance(i).stream().filter(t -> env.taskToStop(t)))
+            .collect(Collectors.toList());
+
+    List<SchedulingAction> stopTaskActions =
+        tasksToStop.stream().map(t -> env.stopTaskFor(t)).collect(Collectors.toList());
+
+    List<SchedulingAction> startTaskActions =
+        summary
+            .getInstances()
+            .stream()
+            .filter(i -> env.needsToAssign(summary.tasksForInstance(i)))
+            .filter(i -> env.noRunningTasks(summary.tasksForInstance(i)))
+            .map(i -> env.startTaskFor(i))
+            .collect(Collectors.toList());
+
+    startTaskActions.addAll(stopTaskActions);
+    return startTaskActions;
```

Comment:
The return tasks are all actions, should be in a better name. or return a list with two actions list like `ImmutableList.Builder().addAll(startTaskActions).addAll(stopTaskActions).build()`

Reply:
Yes, this is clearer!